src/research_index_backend/doi.py

```diff
@@ -0,0 +1,195 @@
+"""DOI (Digital Object Identifier) validation and management module.
+
+This module handles:
+1. DOI pattern validation
+2. Database existence checks
+3. Metadata validation
+4. Batch processing with limits: TODO
+"""
+
+import time
+from logging import getLogger
+from re import IGNORECASE, compile
+from typing import Dict, List
+
+from neo4j import Driver
+from pydantic import BaseModel
+
+from .session import connect_to_db
+
+logger = getLogger(__name__)
+
+# Use regex pattern from
+# https://www.crossref.org/blog/dois-and-matching-regular-expressions/
+DOI_PATTERN = r"10\.\d{4,9}/(?=.*\d)[-._;()/:A-Z0-9]+$"
+
+class DOI(BaseModel):
+    doi: str
+    valid_pattern: bool = False
+    already_exists: bool = False
+    openalex_metadata: bool = False
+    openaire_metadata: bool = False
+    ingestion_success: bool = False
+
+
+class DOITracker(BaseModel):
+    doi_tracker: Dict[str, DOI]
+
+
+class DOIManager:
+    def __init__(
+        self, list_of_dois: List[str], limit: int, update_metadata=True
+    ) -> None:
+
+        if not list_of_dois:
+            raise ValueError("DOI list cannot be empty")
+        if (limit <= 0) or (limit > len(list_of_dois)):
+            raise ValueError(
+                "Limit must be positive and less than the number of DOIs"
+            )
+
+        self.list_of_dois = [
+            doi.strip()
+            .rstrip(".")
+            .replace("https://doi.org/", "")
+            .replace("doi.org/", "")
+            for doi in list_of_dois
+        ]
+        self.limit = limit
+        self.update_metadata = update_metadata
+        self.doi_tracker: DOITracker = {
+            doi: DOI(doi=doi) for doi in self.list_of_dois[: self.limit]
+        }
+        self.PATTERN = compile(DOI_PATTERN, IGNORECASE)
+
+    def start_ingestion(self):
+        self.start_time = time.time()
+
+    def end_ingestion(self):
+        self.end_time = time.time()
+
+    def pattern_check(self):
+        try:
+            self.valid_pattern_dois = []
+            self.invalid_pattern_dois = []
+
+            for doi in self.doi_tracker:
+                if search := self.PATTERN.search(doi):
+                    logger.debug(f"Valid DOI pattern: {search.group()}")
+                    self.doi_tracker[doi].valid_pattern = True
+                    self.valid_pattern_dois.append(doi)
+                else:
+                    logger.warning(f"Invalid DOI pattern: {doi}")
+                    self.invalid_pattern_dois.append(doi)
+            self.num_valid_pattern_dois = len(self.valid_pattern_dois)
+            self.num_invalid_pattern_dois = len(self.invalid_pattern_dois)
+        except Exception as e:
+            logger.error(f"Error whilst checking DOI pattern: {e}")
+            raise
+
+    @connect_to_db
+    def search_dois(self, db: Driver):
+        if not self.valid_pattern_dois:
+            msg = "No DOIs have passed the pattern check and make sure to run pattern check first."
```

Comment:
This part "and make sure to run pattern check first" does not make sense to a user who runs this from the command line as it refer to internal implementation of the code.